File: Excel/700-799/781/781 Challenge.py
import pandas as pd
from itertools import combinations

path = "700-799/781/781 Common Between 2 Columns.xlsx"
input = pd.read_excel(path, usecols="A:C", nrows=12)
test = pd.read_excel(path, usecols="D", nrows=6).sort_values("Answer Expected").reset_index(drop=True)

cols  = [c for c in input.columns if c.startswith("Animals")]
sets  = {c: set(input[c].dropna().unique()) for c in cols}
pairs = {f"{a} & {b}": sorted(sets[a] & sets[b]) for a, b in combinations(cols, 2)}

k = 2
common_k = (
    input[cols].melt(var_name="col", value_name="a")
      .dropna().drop_duplicates(["col","a"])
      .value_counts("a")
      .loc[lambda s: s >= k].index.tolist()
)
common_k.sort()

print(common_k == test["Answer Expected"].to_list()) # True

File: Excel/700-799/781/781 Challenge.R
library(tidyverse)
library(readxl)

path = "Excel/700-799/781/781 Common Between 2 Columns.xlsx"
input = read_excel(path, range = "A1:C12")
test  = read_excel(path, range = "D1:D7") %>% arrange(`Answer Expected`)

k = 2
result = input %>% 
  select(starts_with("Animals")) %>%
  pivot_longer(everything(), values_to = "a") %>%
  filter(!is.na(a)) %>% distinct(name, a) %>% count(a) %>%
  filter(n >= k) %>%
  select(a) %>%
  arrange(a)

all.equal(result$a, test$`Answer Expected`)
# > [1] TRUE